x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java

```diff
@@ -166,24 +153,70 @@ protected void doExecute(Task task, GetDataFrameAnalyticsStatsAction.Request req
-        List<Stats> stoppedTasksStats = stoppedTasksIds.stream().map(this::buildStatsForStoppedTask).collect(Collectors.toList());
-        List<Stats> allTasksStats = new ArrayList<>(runningTasksResponse.getResponse().results());
-        allTasksStats.addAll(stoppedTasksStats);
-        Collections.sort(allTasksStats, Comparator.comparing(Stats::getId));
-        listener.onResponse(new GetDataFrameAnalyticsStatsAction.Response(new QueryPage<>(
-            allTasksStats, allTasksStats.size(), GetDataFrameAnalyticsAction.Response.RESULTS_FIELD)));
+        if (stoppedTasksIds.isEmpty()) {
+            listener.onResponse(runningTasksResponse);
+            return;
+        }
+
+        AtomicInteger counter = new AtomicInteger(stoppedTasksIds.size());
+        AtomicArray<Stats> stoppedStats = new AtomicArray<>(stoppedTasksIds.size());
+        for (int i = 0; i < stoppedTasksIds.size(); i++) {
+            int slot = i;
+            String analyticsId = stoppedTasksIds.get(i);
+            searchStoredProgress(analyticsId, ActionListener.wrap(
+                progress -> {
+                    stoppedStats.set(slot, buildStats(analyticsId, progress));
+                    if (counter.decrementAndGet() == 0) {
+                        List<Stats> allTasksStats = new ArrayList<>(runningTasksResponse.getResponse().results());
+                        allTasksStats.addAll(stoppedStats.asList());
+                        Collections.sort(allTasksStats, Comparator.comparing(Stats::getId));
+                        listener.onResponse(new GetDataFrameAnalyticsStatsAction.Response(new QueryPage<>(
+                            allTasksStats, allTasksStats.size(), GetDataFrameAnalyticsAction.Response.RESULTS_FIELD)));
+                    }
+                },
+                listener::onFailure
+            ));
+        }
     }
 
     static List<String> determineStoppedTasksIds(List<String> expandedIds, List<Stats> runningTasksStats) {
         Set<String> startedTasksIds = runningTasksStats.stream().map(Stats::getId).collect(Collectors.toSet());
         return expandedIds.stream().filter(id -> startedTasksIds.contains(id) == false).collect(Collectors.toList());
     }
 
-    private GetDataFrameAnalyticsStatsAction.Response.Stats buildStatsForStoppedTask(String concreteAnalyticsId) {
-        return buildStats(concreteAnalyticsId, null);
+    private void searchStoredProgress(String configId, ActionListener<List<PhaseProgress>> listener) {
+        SearchRequest searchRequest = new SearchRequest(AnomalyDetectorsIndex.jobStateIndexPattern());
+        searchRequest.indicesOptions(IndicesOptions.lenientExpandOpen());
+        searchRequest.source().size(1);
+        searchRequest.source().query(QueryBuilders.idsQuery().addIds(DataFrameAnalyticsTask.progressDocId(configId)));
+
+        executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, ActionListener.wrap(
```

Comment:
as commented above, I think a single query for all progress docs would be preferable, you should be able to parse them back out correctly given each hit's doc ID.